{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "166369fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Created on Mon Dec  6 16:29:49 2021\n",
    "\n",
    "@author: meraidandouch\n",
    "\"\"\"\n",
    "import numpy as np \n",
    "\n",
    "import scipy.stats as sps\n",
    "\n",
    "def read_data(file_name):\n",
    "    \"\"\"\n",
    "    This file reads the fasta file with all rbd sequences\n",
    "    \n",
    "    file_name: rbd.fasta, a fasta file containing aa sequences of rbd sites in the COVID family \n",
    "    \n",
    "    return data: a dictionary, accesion number ids with associated aa sequence values \n",
    "    \"\"\"\n",
    "    data = {}\n",
    "    with open(file_name, 'rt') as f: \n",
    "        read_data = f.readline().strip('\\n')\n",
    "        while read_data != '':\n",
    "            if read_data[0] == '>': \n",
    "                f_id = read_data\n",
    "            else: \n",
    "                if f_id not in data.keys():\n",
    "                    data[f_id] = read_data\n",
    "                else: \n",
    "                    data[f_id] += read_data\n",
    "            read_data = f.readline().strip('\\n')\n",
    "    return data\n",
    "def count_aa(data): \n",
    "    \"\"\"\n",
    "    This functions concatenates all sequences counts the number of amino acids\n",
    "    \n",
    "    data: a dictionary, accesion number ids with associated aa sequence values \n",
    "    \n",
    "    returns aa_dict, aa ids with associated sequence values  \n",
    "    returns master_count, total number of aa in all sequences \n",
    "    \"\"\"\n",
    "    master_seq = \"\"\n",
    "    master_count = 0\n",
    "    aa = \"QEFLIPGATCNURYSWKMBDHVX\"\n",
    "    aa_dict = {}\n",
    "    for keys, values in data.items():\n",
    "        master_seq += values\n",
    "    for char in master_seq:\n",
    "        if char in aa: \n",
    "            master_count += 1\n",
    "            if char not in aa_dict.keys():\n",
    "                aa_dict[char] = 1\n",
    "            else:\n",
    "                aa_dict[char] += 1\n",
    "    return aa_dict, master_count\n",
    "\n",
    "def freq_aa(aa_dict, master_count):\n",
    "    \"\"\"\n",
    "    The function takes the frequency of each amino acid by dividing the number \n",
    "    of aa by the total aa\n",
    "    \n",
    "    aa_dict: aa ids with associated sequence values  \n",
    "    master_count:  total number of aa in all sequences \n",
    "    \n",
    "    returns freq_dict, aa as keys and freq as associated values \n",
    "    \"\"\"\n",
    "    mc = master_count\n",
    "    freq_dict = {}\n",
    "    a = 0 \n",
    "    for keys, values in aa_dict.items(): \n",
    "            freq_dict[keys] = values/mc\n",
    "            a += (values/mc)\n",
    "    return freq_dict\n",
    "      \n",
    "class base_walker: \n",
    "    def __init__(self):\n",
    "        \"\"\"\n",
    "        this class is the parent class base_walker \n",
    "        and generates aa sequences per random_walker object \n",
    "        \"\"\"\n",
    "        pass\n",
    "    def one_step(self):\n",
    "        pass\n",
    "    def steps(self, seq):\n",
    "        \"\"\"\n",
    "        seq : aa sequence from previous iteration \n",
    "          \n",
    "        return none\n",
    "        \"\"\"\n",
    "        for char in seq:\n",
    "            self.one_step(char)\n",
    "            \n",
    "# one solution:\n",
    "class random_walker(base_walker):\n",
    "    def __init__(self, freq_dict):\n",
    "        \"\"\"\n",
    "        this class inherits the abstract parent class base_walker \n",
    "        and generates aa per random_walker object \n",
    "        \"\"\"\n",
    "        # define stored variables, initial data\n",
    "        self.x = \"\" #step, assumes that x = 0 at the start of the random walk.\n",
    "        self.walk = self.x #list of random steps in a walk \n",
    "        self.prob = freq_dict \n",
    "        \n",
    "    def one_step(self, aa):\n",
    "        \"\"\"\n",
    "        this function determines whether a substitution events will \n",
    "        occur and if it does it generates a random aa else returns \n",
    "        the original aa, it then appends the aa to the remaining current\n",
    "        sequence\n",
    "        \n",
    "        Return None\n",
    "        \"\"\"\n",
    "        temp_dict = self.prob #store temp dictionary, keep original unchanged\n",
    "      \n",
    "        new_keys = list(temp_dict.keys()) #store keys\n",
    "        new_values = list(temp_dict.values())#store values\n",
    "        \n",
    "        event = np.random.choice(['s', 'c'],p=[0.24,0.76]) #(Shah et al. 2020) mutation prob\n",
    "        if event == 's' and aa != \"[YRNPFEDIS]\": #if a substituition event\n",
    "            if aa == 'Y': #(Jia and Gong 2021) mutation prob\n",
    "                aa = 'F'\n",
    "            elif aa == 'S':#(Jia and Gong 2021) mutation prob\n",
    "                aa = 'N'\n",
    "            else: \n",
    "                self.x = np.random.choice(new_keys, p=new_values) #use freq of aa as probabilties \n",
    "        elif event == 'c': #if no substitution event, then return same aa \n",
    "            self.x = aa\n",
    "            \n",
    "        self.walk += self.x\n",
    "    def get_walk(self):\n",
    "        \"\"\"\n",
    "        Returns a list of the positions(scoring) of one walk \n",
    "        \"\"\"\n",
    "        # returns the history of the position of the walker\n",
    "        return self.walk     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a1049308",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">UFP04971.1[OMICRON]: NKLDSKVSGNYNYLYRLFRKSNLKPFERDISTEIYQAGNKPCNGVAGFNCYFPLRSYSFRPTYGVGHQPY\n",
      "\n",
      "New Mutated Sequence:  MKLFSKVSGNYNNSSRLFVEEYLKPFERDISTWIYQAGNNPCNGVAGENCCFYLRSSSFRPTYTKGHQPY\n"
     ]
    }
   ],
   "source": [
    "file_name = 'rbm (1).fasta'\n",
    "data = read_data(file_name)\n",
    "aa_dict, mc = count_aa(data)\n",
    "freq_dict = freq_aa(aa_dict, mc)\n",
    "total_seqs = [] \n",
    "total_dict = {}\n",
    "\n",
    "my_walker = random_walker(freq_dict)\n",
    "my_walker.steps(data['>UFP04971.1 surface glycoprotein [omicron variant Severe acute respiratory syndrome coronavirus 2]'])\n",
    "new_walk = my_walker.get_walk()\n",
    "\n",
    "print('>UFP04971.1[OMICRON]:',data['>UFP04971.1 surface glycoprotein [omicron variant Severe acute respiratory syndrome coronavirus 2]'])\n",
    "print()\n",
    "print(\"New Mutated Sequence: \", new_walk)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e989623f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
